### PR TITLE
RDM-12490 prune result sets

### PR DESCRIPTION
**Pull request by RebeccaBaker**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-12490

### Change description ###
prune result sets based on security classification and crud privileges 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/v2/internal/controller/UIGlobalSearchTempController.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
+package uk.gov.hmcts.ccd.v2.internal.controller;
```
Comment (RebeccaBaker) [inline src/main/java/uk/gov/hmcts/ccd/v2/internal/controller/UIGlobalSearchTempController.java:1]: temp controller - to be removed once testing is completed

File: src/main/java/uk/gov/hmcts/ccd/domain/model/search/global/SearchCriteriaResponseTEMP.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.hmcts.ccd.domain.model.search.global;
```
Comment (RebeccaBaker) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/search/global/SearchCriteriaResponseTEMP.java:1]: temp classes for the controller use only - to be removed once testing is complete

File: build.gradle [hunk 3]
```diff
@@ -572,7 +572,7 @@ jsonSchema2Pojo {
 
 sonarqube {
     properties {
-        property "sonar.exclusions", "build/generated-sources/**/*.java,**/AuditRepository.java,**/CaseDetailsEndpoint.java,**/MidEventCallback.java,**/UIDefinitionRepository.java,**/CachedUIDefinitionGateway.java,**/CaseAccessEndpoint.java,**/UICaseSearchController.java,**/CaseSearchResultViewGenerator.java,**/CaseTypeService.java,**/DefaultCaseDefinitionRepository.java"
+        property "sonar.exclusions", "build/generated-sources/**/*.java,**/AuditRepository.java,**/CaseDetailsEndpoint.java,**/MidEventCallback.java,**/UIDefinitionRepository.java,**/CachedUIDefinitionGateway.java,**/CaseAccessEndpoint.java,**/UICaseSearchController.java,**/CaseSearchResultViewGenerator.java,**/CaseTypeService.java,**/DefaultCaseDefinitionRepository.java, **/SearchCriteriaResponseTEMP.java, **/UIGlobalSearchTempController.java"
```
Comment (RebeccaBaker) [inline build.gradle:575]: added temp classes here as they will be removed after testing to get accurate code coverage

File: src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Autowired
+    public GlobalSearchParser(@Qualifier(CachedUserRepository.QUALIFIER)
+                                  UserRepository userRepository,
+                              CaseTypeService caseTypeService) {
+        this.userRepository = userRepository;
+        this.caseTypeService = caseTypeService;
+    }
+
+    public List<SearchCriteriaResponse> filterCases(List<SearchCriteriaResponse> results, SearchCriteria request) {
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java:39]:
```suggestion
    public List<SearchCriteriaResponse> filterCases(List<CaseDetails> results, SearchCriteria request) {
```

File: src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            Optional<CaseFieldDefinition> caseFieldDefinition =
+                (field.contains(FIELD_SEPARATOR)) ? caseTypeDefinition.getComplexSubfieldDefinitionByPath(field)
+                    : caseTypeDefinition.getCaseField(field);
+            if (caseFieldDefinition.isPresent() && (!AccessControlService
+                .hasAccessControlList(userRepository.getUserRoles(), CAN_READ,
+                    caseFieldDefinition.get().getAccessControlLists()))){
+                condition = false;
+                break;
+            }
+            else if (!securityClassificationService
```
Comment (RebeccaBaker) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java:65]: will be one if statement - for debugging of unit tests - have split out

File: src/main/java/uk/gov/hmcts/ccd/domain/model/search/global/SearchCriteria.java [hunk 6]
```diff
@@ -35,4 +35,27 @@
     @ValidPartiesItems
     @Valid
     private List<Party> parties;
+
+    public enum SearchCriteriaEnum {
+        REGION("caseManagementRegionIds", "CaseLocation.region"),
+        BASE_LOCATION("caseManagementBaseLocationIds", "CaseLocation.baseLocation"),
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/search/global/SearchCriteria.java:41]:
```suggestion
        REGION("caseManagementRegionIds", "caseManagementLocation.region"),
        BASE_LOCATION("caseManagementBaseLocationIds", "caseManagementLocation.baseLocation"),
```

File: src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    public List<CaseDetails> filterCases(List<CaseDetails> results, SearchCriteria request) {
+        List<String> fields = findFieldsToFilter(request);
+        results.removeIf(caseDetails -> !authorised(fields, caseDetails));
+
+        return results;
+    }
+
+    private boolean authorised(List<String> fields, CaseDetails caseDetails) {
+        CaseTypeDefinition caseTypeDefinition = caseTypeService.getCaseType(caseDetails.getCaseTypeId());
+        boolean condition = true;
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java:52]: Just a more meaningful name
```suggestion
        boolean isAuthorised = true;
```

File: src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                break;
+            }
+
+        }
+        return condition;
+    }
+
+    private List<String> findFieldsToFilter(SearchCriteria request) {
+        List<String> fields = new ArrayList<>();
+        if (request.getCaseManagementBaseLocationIds() != null) {
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java:78]:
```suggestion
        if (!CollectionUtils.isEmpty(request.getCaseManagementBaseLocationIds())) {
```
Think it's worth adding a check for not empty too (this does both)

File: src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java (hunk 7 above)
Comment (danlysiak) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java:52]: Actually may just be able to delete this based on next comments!

File: src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            securityClassificationService.getUserClassification(caseDetails.getJurisdiction());
+        for (String field : fields) {
+            Optional<CaseFieldDefinition> caseFieldDefinition =
+                (field.contains(FIELD_SEPARATOR)) ? caseTypeDefinition.getComplexSubfieldDefinitionByPath(field)
+                    : caseTypeDefinition.getCaseField(field);
+            if (caseFieldDefinition.isPresent() && (!AccessControlService
+                .hasAccessControlList(userRepository.getUserRoles(), CAN_READ,
+                    caseFieldDefinition.get().getAccessControlLists()))){
+                condition = false;
+                break;
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java:63]:
```suggestion
                return false;
```
Do we need to `break`? If there's no processing in this method afterwards then we can just return it here.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            }
+            else if (!securityClassificationService
+                .userHasEnoughSecurityClassificationForField(caseTypeDefinition.getJurisdictionId(),
+                    caseTypeDefinition, caseFieldDefinition.get().getId())) {
+                condition = false;
+                break;
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java:69]:
```suggestion
                return false;
```

File: src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            }
+
+        }
+        return condition;
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/globalsearch/GlobalSearchParser.java:73]:
```suggestion
        return true;
```

**Review by danlysiak — APPROVED** [review]

**Pull request merged**